File: tsp-java/tsp-core/src/main/java/dev/tsp/core/IRGenerator.java
package dev.tsp.core;

import dev.tsp.core.model.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class IRGenerator {

    public static TIRResult generate(
        List<ASTNode> ast, Map<String, Object> data, String schemaId, String templateId) {
        List<TIRNode> nodes = new ArrayList<>(ast.size());
        for (ASTNode n : ast) nodes.add(nodeToTir(n, data));
        return new TIRResult(templateId, schemaId, nodes);
    }

    private static TIRNode nodeToTir(ASTNode node, Map<String, Object> data) {
        return switch (node) {
            case TextNode t -> new TIRTextNode(t.content());
            case ExprNode e -> new TIRExprNode(e.field(), resolve(data, e.field()));
            case IfNode i -> {
                boolean cond = evaluate(i.condition(), data);
                List<ASTNode> branch = cond ? i.thenBranch() : i.elseBranch();
                List<TIRNode> out = new ArrayList<>();
                for (ASTNode n : branch) out.add(nodeToTir(n, data));
                yield new TIRIfNode(cond, out);
            }
            case ForEachNode f -> {
                Object iter = resolve(data, f.iterable());
                List<?> items = iter instanceof List<?> list ? list : List.of();
                List<List<TIRNode>> rendered = new ArrayList<>();
                for (Object item : items) {
                    Map<String, Object> scope = new HashMap<>(data);
                    scope.put(f.varName(), item);
                    List<TIRNode> body = new ArrayList<>();
                    for (ASTNode n : f.body()) body.add(nodeToTir(n, scope));
                    rendered.add(body);
                }
                yield new TIRForeachNode(f.varName(), rendered);
            }
        };
    }

    @SuppressWarnings("unchecked")
    static Object resolve(Map<String, Object> data, String path) {
        Object current = data;
        for (String part : path.split("\\.")) {
            if (!(current instanceof Map<?, ?> m) || !m.containsKey(part)) return null;
            current = ((Map<String, Object>) m).get(part);
        }
        return current;
    }

    static boolean evaluate(Condition c, Map<String, Object> data) {
        if ("==".equals(c.op())) {
            Object left = resolve(data, c.left());
            return String.valueOf(left).equals(String.valueOf(c.right()));
        }
        return false;
    }

    private IRGenerator() {}
}

File: templane-python/src/templane_core/ir_generator.py
from __future__ import annotations
from typing import Any
from .models import (
    ASTNode, TextNode, ExprNode, IfNode, ForEachNode, Condition,
    TIRNode, TIRTextNode, TIRExprNode, TIRIfNode, TIRForeachNode, TIRResult,
)


def _resolve(data: dict, path: str) -> Any:
    current: Any = data
    for part in path.split("."):
        if not isinstance(current, dict) or part not in current:
            return None
        current = current[part]
    return current


def _evaluate(condition: Condition, data: dict) -> bool:
    if condition.op == "==":
        return str(_resolve(data, condition.left)) == str(condition.right)
    return False


def generate(
    ast_nodes: list[ASTNode],
    data: dict,
    schema_id: str,
    template_id: str,
) -> TIRResult:
    return TIRResult(
        template_id=template_id,
        schema_id=schema_id,
        nodes=[_node(n, data) for n in ast_nodes],
    )


def _node(node: ASTNode, data: dict) -> TIRNode:
    if isinstance(node, TextNode):
        return TIRTextNode(content=node.content)

    if isinstance(node, ExprNode):
        return TIRExprNode(field=node.field, resolved=_resolve(data, node.field))

    if isinstance(node, IfNode):
        cond = _evaluate(node.condition, data)
        branch = node.then_branch if cond else node.else_branch
        return TIRIfNode(condition=cond, branch=[_node(n, data) for n in branch])

    if isinstance(node, ForEachNode):
        items_val = _resolve(data, node.iterable)
        items: list[Any] = items_val if isinstance(items_val, list) else []
        rendered = []
        for item in items:
            scope = {**data, node.var: item}
            rendered.append([_node(n, scope) for n in node.body])
        return TIRForeachNode(var=node.var, items=rendered)

    raise ValueError(f"Unknown AST node: {node}")
